test/python/utils/test_deprecation.py

```diff
@@ -0,0 +1,465 @@
+                  Stop using my_func!
+                """
+            ),
+        )
+
+    def test_deprecate_arguments_runtime_warning(self) -> None:
+        """Test that `@deprecate_arguments` warns whenever the arguments are used.
+
+        Also check that old arguments are passed in as their new alias.
+        """
+
+        @deprecate_arguments({"arg1": None, "arg2": "new_arg2"}, since="9.99")
+        def my_func(*, arg1: str = "a", new_arg2: str) -> None:
+            del arg1
+            self.assertEqual(new_arg2, "z")
+
+        my_func(new_arg2="z")  # No warnings if no deprecated args used.
+        with self.assertWarnsRegex(DeprecationWarning, "arg1"):
+            my_func(arg1="a", new_arg2="z")
+        with self.assertWarnsRegex(DeprecationWarning, "arg2"):
+            # `arg2` should be converted into `new_arg2`.
+            my_func(arg2="z")  # pylint: disable=missing-kwoa
+
+    def test_deprecate_function_runtime_warning(self) -> None:
+        """Test that `@deprecate_function` warns whenever the function is used."""
+
+        @deprecate_function("Stop using my_func!", since="9.99")
+        def my_func() -> None:
+            pass
+
+        with self.assertWarnsRegex(DeprecationWarning, "Stop using my_func!"):
+            my_func()
+
+
+class DeprecationExtensionTest(QiskitTestCase):
+    """Test that we correctly insert the deprecation directive at the right location.
+
+    These test cases were manually created by changing the docstring of a function in
+    ``qiskit-ibmq-provider`` and building the docs to ensure that everything rendered correctly.
+    (Any Qiskit repo will do.) To get the ``expected`` lines, you can add this extension to
+    ``conf.py``:
+
+        def print_docstring(app, what, name, obj, options, lines):
+            # INSTRUCTIONS: Change the below value to the function you're testing with it.
+            if name != "qiskit.providers.ibmq.credentials.Credentials.connection_parameters":
+                return
+            print("HERE")  # Useful so that you can grep for "HERE" in the Sphinx output.
+            print(obj.__doc__)
+
+        def setup(app):
+            app.connect('autodoc-process-docstring', print_docstring)
+
+    Then, use the function you set with the `name !=` line, and modify its docstring. Build the
+    docs and search for "HERE" to see the relevant output.
+    """
+
+    def assert_docstring(
+        self,
+        func: Callable,
+        *,
+        expected: str,
+        msg: str = "Deprecated!",
+        pending: bool = False,
+        since_is_set: bool = True,
+        num_deprecations: int = 1,
+    ) -> None:
+        """Add docstring to ``func`` and check that it worked as expected."""
+        for _ in range(num_deprecations):
+            _add_deprecation_to_docstring(
+                func, msg=msg, since="9.99" if since_is_set else None, pending=pending
+            )
+        self.assertEqual(func.__doc__, expected)
```

Comment:
The way this is currently written makes the tests harder to read to me; a fair amount of logic of "what is being tested?" actually lives in this assertion function (`num_deprecations`, `since_is_set`), and it's inherently coupled to the test sites because it defines magic values for the expected output (the values of `msg`, `since` and `pending`).

I think the tests would be clearer without this indirection - for example, the test immediately below this one would just be
```python
def test_deprecation_docstring_added_to_end_when_no_meta_lines(self):
    @deprecate_function("Deprecated!", since=9.99)
    def func1():
        pass

    self.assertEqual(func1.__doc__, dedent(""" ... """))
```
and the ones about message repetition would be a direct test of `deprecate_arguments`. We end up repeating a couple of the arguments, but we have to do that anyway, because they're already in the `expected` output.  I also don't think a little bit of verbosity is necessarily bad in a test suite.

Secondly: I'm of the school that private functions should be implementation details, and the vast majority (if not all) of the unit tests should only be of the public interface.  I get that that's not universally agreed on, especially because Python allows us to ignore the "access modifiers", but it's more in keeping with most of Qiskit's tests, and in the cases where it hasn't been followed here, I've been burned by having to change tests during bugfixes for public behaviour that hadn't changed.  A variant literally just happened to me a few minutes ago on #9726.

Reply:
I got rid of the test helper function. I could go either way, I'm not very wedded to it in this particular case.

--

Re private function, your comment made me realize that `add_deprecation_to_docstring` should be a public function. While my hope is non-Terra wants to use our new deprecation helpers, we are not forcing them to. It's helpful to let them keep using their helpers but now call this function `add_deprecation_to_docstring` inside their helpers.

Thanks!